src/EFCore.PG/Query/ExpressionTranslators/Internal/NpgsqlArrayMethodCallTranslator.cs

```diff
@@ -0,0 +1,401 @@
+                return VersionAtLeast(9, 1)
+                    ? new SqlFunctionExpression("string_to_array", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(3))
+                    : VersionAtLeast(7, 4)
+                        ? new SqlFunctionExpression("string_to_array", expression.Method.ReturnType, expression.Arguments.Skip(1).Take(2))
+                        : throw VersionNotSupportedException(nameof(NpgsqlArrayExtensions.StringToList), 7, 4);
+
+            default:
+                return null;
+            }
+
+            NotSupportedException VersionNotSupportedException(string name, int major, int minor)
+                => new NotSupportedException($"{nameof(NpgsqlArrayExtensions)}.{name} is not supported before PostgreSQL {major}.{minor}.");
+        }
+
+        /// <summary>
+        /// Translates static methods defined on <see cref="Array"/>.
+        /// </summary>
+        /// <param name="expression">The expression to translate.</param>
+        /// <returns>
+        /// A translated expression or null if no translation is supported.
+        /// </returns>
+        [CanBeNull]
+        Expression ArrayStaticHandler([NotNull] MethodCallExpression expression)
+        {
+            if (!expression.Method.IsStatic || expression.Method.DeclaringType != typeof(Array))
+                return null;
+
+            switch (expression.Method.Name)
+            {
+            case nameof(Array.IndexOf) when VersionAtLeast(9, 5) && expression.Method.GetParameters().Length <= 3:
+                return Coalesce(new SqlFunctionExpression("array_position", typeof(int), expression.Arguments.Take(3)));
+
+            default:
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Translates instance methods defined on <see cref="Array"/>.
+        /// </summary>
+        /// <param name="expression">The expression to translate.</param>
+        /// <returns>
+        /// A translated expression or null if no translation is supported.
+        /// </returns>
+        [CanBeNull]
+        Expression ArrayInstanceHandler([NotNull] MethodCallExpression expression)
+        {
+            if (!(expression.Object is Expression instance))
+                return null;
+
+            if (!instance.Type.IsArray)
+                return null;
+
+            switch (expression.Method.Name)
+            {
+            case nameof(Array.GetLength) when VersionAtLeast(8, 4):
+                return
+                    new SqlFunctionExpression(
+                        "array_length",
+                        typeof(int),
+                        new[] { instance, GenerateOneBasedIndexExpression(expression.Arguments[0]) });
+
+            case nameof(Array.GetLowerBound) when VersionAtLeast(7, 4):
+                return
+                    Coalesce(
+                        new SqlFunctionExpression(
+                            "array_lower",
```

Comment:
PostgreSQL arrays are 1-based by default unlike .NET arrays which are 0-based by default. So the lower (and upper) bounds have to be adjusted.

Also, why are we coalescing? In what situation will a null be returned?

Reply:
Ah, I didn't think about needing to adjust incoming (just outgoing)! I'll fix that.

As for the coalesce: 

```
postgres=# SELECT array_lower(ARRAY[]::int[], 1) IS NULL;
 ?column?
----------
 t
(1 row)
```

Whereas the .NET functions are [documented (old docs)](https://msdn.microsoft.com/en-us/library/system.array.getlowerbound(v=vs.110).aspx) to return `-1` when empty. The [new docs](https://docs.microsoft.com/en-us/dotnet/api/system.array.getlowerbound?view=netframework-4.7.2) indicate that a value is always returned, though they no longer list `-1` as the returned value.

Reply:
Ok, makes sense then (maybe worth confirming the .NET behavior).